File: Java/examples/uuid_utils_example.java
package examples;

import uuid_utils.mod;

/**
 * UUID Utilities Example
 *
 * Demonstrates usage of UUID generation and manipulation functions.
 *
 * Compile: javac -cp .. examples/uuid_utils_example.java
 * Run: java -cp .. examples.uuid_utils_example
 */
public class uuid_utils_example {

    public static void main(String[] args) {
        System.out.println("=== UUID Utilities Examples ===\n");

        exampleV4Generation();
        exampleV3V5Generation();
        exampleRandomStrings();
        examplePasswordGeneration();
        exampleValidation();
        exampleFormatConversion();
        exampleAlternativeIds();

        System.out.println("\n=== All examples completed ===");
    }

    private static void exampleV4Generation() {
        System.out.println("1. UUID v4 (Random) Generation");
        System.out.println("------------------------------");

        // Standard format with dashes
        String uuid1 = mod.generateV4();
        System.out.println("Standard v4:  " + uuid1);

        // Compact format without dashes
        String uuid2 = mod.generateV4Compact();
        System.out.println("Compact v4:   " + uuid2);

        // Uppercase format
        String uuid3 = mod.generateV4Upper();
        System.out.println("Uppercase v4: " + uuid3);

        System.out.println();
    }

    private static void exampleV3V5Generation() {
        System.out.println("2. UUID v3/v5 (Name-based) Generation");
        System.out.println("--------------------------------------");

        // v3 uses MD5 - deterministic
        String v3_1 = mod.generateV3(mod.NAMESPACE_DNS, "example.com");
        String v3_2 = mod.generateV3(mod.NAMESPACE_DNS, "example.com");
        System.out.println("v3 for 'example.com' (DNS): " + v3_1);
        System.out.println("v3 again (same):            " + v3_2);
        System.out.println("Same input = same UUID: " + v3_1.equals(v3_2));

        // v5 uses SHA1 - also deterministic, preferred over v3
        String v5_1 = mod.generateV5(mod.NAMESPACE_URL, "https://example.com");
        String v5_2 = mod.generateV5(mod.NAMESPACE_URL, "https://example.com");
        System.out.println("\nv5 for 'https://example.com' (URL): " + v5_1);
        System.out.println("v5 again (same):                    " + v5_2);

        // Different namespaces produce different UUIDs
        String v5_dns = mod.generateV5(mod.NAMESPACE_DNS, "example.com");
        String v5_url = mod.generateV5(mod.NAMESPACE_URL, "example.com");
        System.out.println("\nSame name, different namespaces:");
        System.out.println("  DNS namespace: " + v5_dns);
        System.out.println("  URL namespace: " + v5_url);

        System.out.println();
    }

    private static void exampleRandomStrings() {
        System.out.println("3. Random String Generation");
        System.out.println("----------------------------");

        // Alphanumeric
        String alpha = mod.randomAlphanumeric(16);
        System.out.println("Alphanumeric (16): " + alpha);

        // Numeric only
        String numeric = mod.randomNumeric(10);
        System.out.println("Numeric (10):      " + numeric);

        // Hexadecimal
        String hex = mod.randomHex(32);
        System.out.println("Hex (32):          " + hex);

        // Custom charset
        String custom = mod.randomString(20, "ABCDEF123456");
        System.out.println("Custom charset:    " + custom);

        System.out.println();
    }

    private static void examplePasswordGeneration() {
        System.out.println("4. Secure Password Generation");
        System.out.println("------------------------------");

        // Generate secure passwords
        String pwd1 = mod.randomPassword(12);
        String pwd2 = mod.randomPassword(16);
        String pwd3 = mod.randomPassword(20);

        System.out.println("Password (12 chars): " + pwd1);
        System.out.println("Password (16 chars): " + pwd2);
        System.out.println("Password (20 chars): " + pwd3);

        System.out.println("\nAll passwords contain:");
        System.out.println("  - Uppercase letters");
        System.out.println("  - Lowercase letters");
        System.out.println("  - Digits");
        System.out.println("  - Special characters");

        System.out.println();
    }

    private static void exampleValidation() {
        System.out.println("5. UUID Validation");
        System.out.println("-------------------");

        String validV4 = mod.generateV4();
        String validV5 = mod.generateV5(mod.NAMESPACE_DNS, "test");
        String invalid = "not-a-uuid";

        System.out.println("Valid v4 UUID: " + validV4);
        System.out.println("  isValidUUID: " + mod.isValidUUID(validV4));
        System.out.println("  isValidV4:   " + mod.isValidV4(validV4));
        System.out.println("  Version:     " + mod.getVersion(validV4));

        System.out.println("\nValid v5 UUID: " + validV5);
        System.out.println("  isValidUUID: " + mod.isValidUUID(validV5));
        System.out.println("  isValidV4:   " + mod.isValidV4(validV5));
        System.out.println("  Version:     " + mod.getVersion(validV5));

        System.out.println("\nInvalid: '" + invalid + "'");
        System.out.println("  isValidUUID: " + mod.isValidUUID(invalid));

        System.out.println();
    }

    private static void exampleFormatConversion() {
        System.out.println("6. Format Conversion");
        System.out.println("---------------------");

        String standard = mod.generateV4();
        System.out.println("Original (standard): " + standard);

        // To compact
        String compact = mod.toCompact(standard);
        System.out.println("To compact:          " + compact);

        // Back to standard
        String backToStandard = mod.toStandard(compact);
        System.out.println("Back to standard:    " + backToStandard);

        // To uppercase
        String upper = mod.toUpperCase(standard);
        System.out.println("To uppercase:        " + upper);

        // Parse to UUID object
        java.util.UUID uuid = mod.parse(standard);
        System.out.println("\nParsed to UUID object: " + uuid);

        System.out.println();
    }

    private static void exampleAlternativeIds() {
        System.out.println("7. Alternative ID Formats");
        System.out.println("--------------------------");

        // Short ID (Base32 encoded)
        String shortId = mod.shortId(8);
        System.out.println("Short ID (8 bytes):  " + shortId);

        // Nano ID (URL-safe, 21 chars)
        String nanoId = mod.nanoId();
        System.out.println("Nano ID (default):   " + nanoId);

        // Custom length Nano ID
        String nanoId10 = mod.nanoId(10);
        System.out.println("Nano ID (10 chars):  " + nanoId10);

        // ULID (sortable, 26 chars)
        String ulid1 = mod.ulid();
        String ulid2 = mod.ulid();
        System.out.println("ULID 1:              " + ulid1);
        System.out.println("ULID 2:              " + ulid2);

        System.out.println();
    }
}

File: TypeScript/examples/uuid_utils_example.ts
/**
 * UUID Utilities Example
 * 
 * Demonstrates usage of the UUID utilities module for TypeScript.
 * 
 * Run with: npx ts-node uuid_utils_example.ts
 */

import {
  v4,
  v1,
  compact,
  uppercase,
  isValid,
  parse,
  getVersion,
  getVariant,
  toCompact,
  toStandard,
  equals,
  nanoId,
  shortId,
  randomString,
  randomPassword,
  UuidUtils,
  CHARSETS,
  UUID_NAMESPACES,
} from '../uuid_utils/mod';

console.log('========================================');
console.log('UUID Utilities Examples');
console.log('========================================\n');

// ==================== UUID Generation ====================

console.log('--- UUID v4 (Random-based) ---');
const uuid4 = v4();
console.log(`Generated UUID v4: ${uuid4}`);
console.log(`Is valid: ${isValid(uuid4)}`);
console.log(`Version: ${getVersion(uuid4)}`);
console.log(`Variant: ${getVariant(uuid4)}`);
console.log();

console.log('--- UUID v1 (Timestamp-based) ---');
const uuid1 = v1();
console.log(`Generated UUID v1: ${uuid1}`);
console.log(`Is valid: ${isValid(uuid1)}`);
console.log(`Version: ${getVersion(uuid1)}`);
console.log();

// ==================== Format Variations ====================

console.log('--- Format Variations ---');
const standard = v4();
console.log(`Standard format:  ${standard}`);
console.log(`Compact format:   ${compact()}`);
console.log(`Uppercase format: ${uppercase()}`);
console.log();

// ==================== Parsing UUIDs ====================

console.log('--- Parsing UUIDs ---');
const parsedV4 = parse(standard);
console.log(`Parsed UUID v4:`, parsedV4);

const parsedV1 = parse(uuid1);
console.log(`Parsed UUID v1:`, parsedV1);
console.log(`  Timestamp: ${parsedV1?.timestamp?.toISOString()}`);
console.log(`  Clock Seq: ${parsedV1?.clockSeq}`);
console.log(`  Node: ${parsedV1?.node}`);
console.log();

// ==================== Validation ====================

console.log('--- Validation ---');
console.log(`Is '550e8400-e29b-41d4-a716-446655440000' valid? ${isValid('550e8400-e29b-41d4-a716-446655440000')}`);
console.log(`Is 'invalid-uuid' valid? ${isValid('invalid-uuid')}`);
console.log(`Is '${standard}' a v4? ${isValid(standard, 4)}`);
console.log(`Is '${uuid1}' a v1? ${isValid(uuid1, 1)}`);
console.log();

// ==================== Format Conversion ====================

console.log('--- Format Conversion ---');
const testUuid = '550e8400-e29b-41d4-a716-446655440000';
const compactUuid = toCompact(testUuid);
const backToStandard = toStandard(compactUuid || '');

console.log(`Original:      ${testUuid}`);
console.log(`To compact:    ${compactUuid}`);
console.log(`Back to standard: ${backToStandard}`);
console.log();

// ==================== UUID Comparison ====================

console.log('--- UUID Comparison ---');
const uuidA = '550e8400-e29b-41d4-a716-446655440000';
const uuidB = '550E8400-E29B-41D4-A716-446655440000';
const uuidC = '6ba7b810-9dad-11d1-80b4-00c04fd430c8';

console.log(`UUID A: ${uuidA}`);
console.log(`UUID B: ${uuidB}`);
console.log(`UUID C: ${uuidC}`);
console.log(`A equals B (case insensitive): ${equals(uuidA, uuidB)}`);
console.log(`A equals C: ${equals(uuidA, uuidC)}`);
console.log();

// ==================== Alternative IDs ====================

console.log('--- Nano ID (URL-safe) ---');
console.log(`Default (21 chars): ${nanoId()}`);
console.log(`Custom (10 chars):  ${nanoId(10)}`);
console.log(`Custom (32 chars):  ${nanoId(32)}`);
console.log();

console.log('--- Short ID (Base62) ---');
console.log(`Default (8 chars):  ${shortId()}`);
console.log(`Custom (16 chars):  ${shortId(16)}`);
console.log();

// ==================== Random String Generation ====================

console.log('--- Random String Generation ---');
console.log(`Alphanumeric (16): ${randomString(16)}`);
console.log(`Digits only (6):   ${randomString(6, CHARSETS.DIGITS)}`);
console.log(`Hex only (16):     ${randomString(16, CHARSETS.HEX)}`);
console.log(`Uppercase (8):     ${randomString(8, CHARSETS.UPPERCASE)}`);
console.log();

// ==================== Secure Password Generation ====================

console.log('--- Secure Password Generation ---');
console.log(`Default (16 chars): ${randomPassword()}`);
console.log(`Short (8 chars):    ${randomPassword(8)}`);
console.log(`Long (32 chars):    ${randomPassword(32)}`);
console.log();

// ==================== Using UuidUtils Namespace ====================

console.log('--- Using UuidUtils Namespace ---');
const nsUuid = UuidUtils.v4();
console.log(`Via namespace: ${nsUuid}`);
console.log(`Is valid: ${UuidUtils.isValid(nsUuid)}`);
console.log();

// ==================== Predefined Namespaces ====================

console.log('--- Predefined Namespaces ---');
console.log(`DNS:  ${UUID_NAMESPACES.DNS}`);
console.log(`URL:  ${UUID_NAMESPACES.URL}`);
console.log(`OID:  ${UUID_NAMESPACES.OID}`);
console.log(`X500: ${UUID_NAMESPACES.X500}`);
console.log();

// ==================== Practical Use Cases ====================

console.log('--- Practical Use Cases ---');

// Generate multiple UUIDs for database records
console.log('\n1. Database Record IDs:');
const records = Array.from({ length: 5 }, (_, i) => ({
  id: v4(),
  name: `Record ${i + 1}`,
}));
records.forEach(record => console.log(`   ${record.name}: ${record.id}`));

// Generate session tokens
console.log('\n2. Session Tokens (Nano ID):');
const sessions = Array.from({ length: 3 }, () => nanoId(32));
sessions.forEach((token, i) => console.log(`   Session ${i + 1}: ${token}`));

// Generate API keys
console.log('\n3. API Keys (Short ID):');
const apiKeys = Array.from({ length: 3 }, () => shortId(24));
apiKeys.forEach((key, i) => console.log(`   Key ${i + 1}: ${key}`));

// Generate temporary passwords
console.log('\n4. Temporary Passwords:');
const passwords = Array.from({ length: 3 }, () => randomPassword(12));
passwords.forEach((pwd, i) => console.log(`   Password ${i + 1}: ${pwd}`));

console.log('\n========================================');
console.log('Examples completed successfully!');
console.log('========================================');